soaktest/soakclient_transaction/soakclient_transaction.go

```diff
@@ -0,0 +1,495 @@
+		return terminate(err, *c)
+	}
+
+	var msgid uint64
+	run := true
+	commited := false
+	for run == true {
+		select {
+		case sig := <-sigchan:
+			fmt.Printf("Caught signal %v: terminating\n", sig)
+			run = false
+			if msgid != 0 && !commited {
+				transactionCommitNum, err = commitTransaction(
+					getAllPartitionPositions(partitionPositionMap),
+					*p,
+					*c,
+					transactionCommitNum)
+			}
+			soaktest.PrintConsumerStatus()
+			return err
+		case <-ticker.C:
+			if msgid != 0 && !commited {
+				transactionCommitNum, err = commitTransaction(
+					getAllPartitionPositions(partitionPositionMap),
+					*p,
+					*c,
+					transactionCommitNum)
+				if err != nil {
+					fmt.Printf("Failed to commit transaction %s\n", err)
+					run = false
+					return terminate(err, *c)
+				}
+				commited = true
+				p.BeginTransaction()
+			}
+
+		default:
+			var ev = c.Poll(100)
+			if ev == nil {
+				continue
+			}
+			switch e := ev.(type) {
+			case *kafka.Message:
+				if !soaktest.HandleMessage(e, hwmarks) {
+					fmt.Printf("Message receive failed, " +
+						"skip producing message")
+					continue
+				}
+				produceMessage(e, *p, msgid, outTopic)
+				msgid += 1
+				position := getConsumerPosition(
+					*c,
+					e.TopicPartition.Partition,
+					inputTopic)
+				partitionPositionMap[e.TopicPartition.Partition] = position
+				if msgid%100 == 0 && !commited {
+					transactionCommitNum, err = commitTransaction(
+						getAllPartitionPositions(partitionPositionMap),
+						*p,
+						*c,
+						transactionCommitNum)
+					if err != nil {
+						fmt.Errorf("Failed to commit transaction %s\n", err)
+						run = false
+						return terminate(err, *c)
+					}
+					commited = true
+					p.BeginTransaction()
+				} else {
+					commited = false
+				}
+
+			case *kafka.Stats:
+				soaktest.HandleStatsEvent(e, inputTopic)
+
+			case kafka.Error:
+				// kafka.Errors should generally be
+				// considered informational, the client
+				// will try to automatically recover.
+				soaktest.ConsumerMsgErrs += 1
+				soaktest.DatadogIncrement(soaktest.CONSUMER_RECEIVE_ERROR_FROM+
+					inputTopic, 1)
+				fmt.Printf("kafka.Error: %v: %v for consumer %v\n",
+					e.Code(), e, c)
+
+			default:
+				fmt.Printf("Ignored %s for consumer %v\n", e, c)
+			}
+		}
+	}
+	return terminate(err, *c)
+}
+
+// Transaction producer, produce messages to output topic
+func produceMessage(e *kafka.Message,
+	p kafka.Producer,
+	msgid uint64,
+	outTopic string) {
+	encodingTime, err := time.Now().GobEncode()
+	if err != nil {
+		fmt.Errorf("Failed to encode current time %s\n", err)
+	}
+
+	err = p.Produce(&kafka.Message{
+		TopicPartition: kafka.TopicPartition{
+			Topic:     &outTopic,
+			Partition: kafka.PartitionAny},
+		Value: []byte(string(e.Value)),
+		Headers: []kafka.Header{{Key: "time",
+			Value: encodingTime},
+			{Key: "msgid",
+				Value: soaktest.ConvertUint64ToByteArray(msgid)}},
+	}, nil)
+	// Wait for delivery report
```

Comment:
This makes producing synchronous, which is an anti-pattern and slow.
Use a separate go-routine or the same loop that calls produceMessage() to read the events channel asynchronously.

See https://github.com/edenhill/librdkafka/wiki/FAQ#why-is-there-no-sync-produce-interface